### DISCUSSION TITLE
ESP32 TouchPad

### DISCUSSION THREAD

**Discussion by krishnak** [micropython/micropython#9622] [Using MicroPython]
How do I trigger an interrupt for the TouchPad on ESP32 based on threshold values?

**Comment by sosi-deadeye** [comment]
There is no method to attach an IRQ to a TouchPad.
Currently, you have to read the value from TouchPad.

If you use asyncio, you could create a Task which reads the TouchPad repeatedly.

```
import uasyncio as asyncio
from machine import Pin, TouchPad


async def async_trigger(tp: TouchPad, threshold, callback):
    while True:
        await asyncio.sleep(0.01)
        if (value := tp.read()) < threshold:
            callback(value)


async def main():
    tp = TouchPad(Pin(32))
    trigger_coro = async_trigger(tp, 400, print)
    asyncio.create_task(trigger_coro)

    while True:
        await asyncio.sleep(1)

asyncio.run(main())
```

Then touch GPIO32 with your finger. You'll get many callbacks (print-function) during the touch.
A synchronous function as callback could block the whole eventloop. This is a potential problem.

Maybe you want to recognize only one touch, then waiting since the value is high enough (no longer touching) and then wait for the next low value (touch). The example code does not memorize the last state.

Example code to with a hysteresis:
```

import uasyncio as asyncio
from machine import Pin, TouchPad


async def async_trigger(
    tp: TouchPad,
    on_threshold: int,
    off_threshold: int,
    callback_on,
    callback_off,
):
    active = False
    while True:
        await asyncio.sleep(0.01)
        
        value = tp.read()
        
        if active and value >= off_threshold:
            active = False
            callback_off(value)
        elif not active and value <= on_threshold:
            active = True
            callback_on(value)


async def main():
    tp = TouchPad(Pin(32))
    trigger_coro = async_trigger(tp, 100, 600, print, print)
    asyncio.create_task(trigger_coro)

    while True:
        await asyncio.sleep(1)

asyncio.run(main())

```

**Comment by krishnak** [comment]
Thank you for your suggestion, I was going to implement it your way of
looping if there was no other method. However I find that there is
something like this

t = TouchPad(Pin(14))t.config(500)               # configure the
threshold at which the pin is considered
touchedesp32.wake_on_touch(True)

Whether this can be some how reused to call the IRQ call back.
